jtag/drivers/bcm2835gpio: Ignore drive settings of input pins on init

bpupp97 opened the pull request:
Input GPIOs were being incorrectly reset to outputs because of faulty logic with push-pull optimizations. This commit will only check pins configured as init-active or init-inactive when initializing push-pull to outputs.

vfazio commented on src/jtag/drivers/bcm2835gpio.c:
```diff
@@ -235,18 +235,21 @@ static void initialize_gpio(enum adapter_gpio_config_index idx)
 
-	/* Direction for non push-pull is already set by set_gpio_value() */
-	if (adapter_gpio_config[idx].drive == ADAPTER_GPIO_DRIVE_MODE_PUSH_PULL)
-		OUT_GPIO(adapter_gpio_config[idx].gpio_num);
```
I don't have a super strong preference, but we could also do:

```c
       if (adapter_gpio_config[idx].drive == ADAPTER_GPIO_DRIVE_MODE_PUSH_PULL &&
           adapter_gpio_config[idx].init_state != ADAPTER_GPIO_INIT_STATE_INPUT)
                OUT_GPIO(adapter_gpio_config[idx].gpio_num);
```
This is the minimum necessary to address the problem and keeps the code structure as is.

It looks like `am335xgpio` driver is also probably affected by this quirk

bpupp97 replied:
Modified my changes to do this instead

vfazio commented:
This patch does not pass `./tools/checkpatch.sh` so please fix that.

vfazio commented:
The commit subject is a little misleading. 
```
jtag/drivers/bcm2835gpio: Don't configure push/pull for inputs on init

Input GPIOs were being incorrectly reset to outputs because of faulty logic
with push-pull optimizations. This commit will only check pins configured as
init-active or init-inactive when initializing push-pull to outputs.
```

We're not configuring whether a pin is push pull when initializing it. Pins seem to default to push/pull regardless of direction and we _are_ configuring input pins during init even if their drive value is `ADAPTER_GPIO_DRIVE_MODE_PUSH_PULL`

It seems like this commit makes sure that pins that are configured as inputs don't get mistakenly flipped to outputs based solely on the value of the pin's drive being `ADAPTER_GPIO_DRIVE_MODE_PUSH_PULL`, which is not an indication of direction.

You should add a commit trailer that calls out it is fixing a previous commit. The trailer should look like:
```
Fixes: 0dd969d83bad ("drivers/bcm2835gpio: Migrate to adapter gpio commands ")
```

vfazio reviewed (changes requested):
Looks ok, please see my comments

bpupp97 committed:
jtag/drivers/bcm2835gpio: Ignore drive settings of input pins on init

Input GPIOs were being incorrectly reset to outputs because of faulty logic
with push-pull optimizations since ADAPTER_GPIO_DRIVE_MODE_PUSH_PULL is the
default drive setting. This commit will only initialize the direction of
push-pull driven pins if they are also not set to initialize as inputs.

Fixes: 0dd969d83bad ("drivers/bcm2835gpio: Migrate to adapter gpio commands ")

Change-Id: I0c6c7436a583b6ae56e4a8d70fa2abed72e6bf09
Signed-off-by: Brandon Pupp <bpupp97@gmail.com>

bpupp97 commented:
Updated title to reflect new commit message

vfazio commented:
This is upstream in https://github.com/openocd-org/openocd/commit/ea0ac6dc779f09fd21e007d93987a6a7eca5149e

closing
